### DISCUSSION TITLE
Using CollectionFixture to Share a DatabaseFixture that requires a shared WebApplicationFactory?

### DISCUSSION THREAD

**Discussion by James-Bryan** [xunit/xunit#2915] [Question]
For the integration tests in my project, I am recording my Test data in a Shared Test Database.
WebApplicationFactory is injected into my Database Fixtures, and is used to share an ApplicationDBContext between the Tests (amongst other things)
The Fixtures use the ApplicationDBContext to seed the Test Database with data, and the saved data is used by my tests.

I have multiple Database Fixtures so I can run a portion of the tests and only need to generate the required test data.
Many of my test classes use the same database fixture via IClassFixture.
This (as expected) causes the fixture constructor to run once for each test file.
As a result, the Test Data is recorded to the Database multiple times. **I am looking to avoid this.**

My current setup:

Main Collection to provide the WebApplicationFactory
```
[CollectionDefinition("Tests")]
public class FixtureCollection : ICollectionFixture<WebApplicationFactory<Startup>>
{

}
```

My Fixtures then use the Injected WebApplicationFactory to
- Get the shared ApplicationDBContext
- Generate HttpClients as required.
- Save test entities, which are used by the tests
```
public class SaveTestDatabaseFixture : WebApplicationFactory<Startup>{
    public SaveTestDatabaseFixture(WebApplicationFactory<Startup> factory)
    {
          var context = factory.GetContext();
          ......
          context.TestData.Add(testData);
          context.SaveChanges();
    }
}
```

My Tests then utilize the Collection and Database Fixture, and retrieve the relevant test data for each test
```
[Collection("Tests")]
public class TestDataTests : IClassFixture<SaveTestDatabaseFixture> {
    private SaveTestDatabaseFixture databaseFixture;

    public TestDataTests (SaveTestDatabaseFixture databaseFixture)
    {
        this.databaseFixture= databaseFixture;
    }

    [Fact]
    public async Task Test1()
    {
        var x = databaseFixture.TestData.Id;
        //Run Test
    }
}

[Collection("Tests")]
public class TestData2Tests : IClassFixture<SaveTestDatabaseFixture> {
    private SaveTestDatabaseFixture databaseFixture;

    public TestData2Tests (SaveTestDatabaseFixture databaseFixture)
    {
        this.databaseFixture= databaseFixture;
    }

    [Fact]
    public async Task Test2()
    {
        var x = databaseFixture.TestData2.Id;
        //Run Test
    }
}
```

As the above tests both use IClassFixture<SaveTestDatabaseFixture>, the Fixture constructor is run twice, and the Test Data is generated twice.

I am looking to generate a shared SaveTestDatabaseFixture, so the test data is generated once, and the Test Files then share the fixture and data.

I have been attempting to do this by implementing SaveTestDatabaseFixture as part of a collection fixture, but I have been unable to identify if I can provide SaveTestDatabaseFixture with the required WebApplicationFactory<Startup>, which by my understanding is being injected via the "Tests" fixture collection above.

```
//[Collection("Tests")]
[CollectionDefinition("SaveTests")]
public class SaveTestCollectionFixture: ICollectionFixture<SaveTestDatabaseFixture>  //,ICollectionFixture<WebApplicationFactory<Startup>>
{

}

[Collection("SaveTests")]
public class TestDataTests {
    private SaveTestDatabaseFixture databaseFixture;

    public TestDataTests (SaveTestDatabaseFixture databaseFixture)
    {
        this.databaseFixture= databaseFixture;
    }

//Tests
}
```

If I try to do something like this, I get the following error.

```
Message: 
System.AggregateException : One or more errors occurred. (Collection fixture type 'SaveTestDatabaseFixture' had one or more unresolved constructor arguments: WebApplicationFactory factory) (The following constructor parameters did not have matching fixture data: SaveTestDatabaseFixture fixture)
---- Collection fixture type 'SaveTestDatabaseFixture' had one or more unresolved constructor arguments: WebApplicationFactory factory
---- The following constructor parameters did not have matching fixture data: SaveTestDatabaseFixture fixture

  Stack Trace: 
----- Inner Stack Trace #1 (Xunit.Sdk.TestClassException) -----
----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----
```

From my understanding, this is because I am not Injecting the WebApplicationFactory into SaveTestDatabaseFixture.

Is it possible to define SaveTestDatabaseFixture as a Collection Fixture, and have WebApplicationFactory<Startup> injected into it?

If not, are there any alternative methods to allow SaveTestDatabaseFixture to be generated once, shared between multiple tests, while having WebApplicationFactory<Startup> injected into it?

If any more information is required, please let me know.

Thanks

**Comment by bradwilson** [comment]
You can't use injection here. You have to create the instance of `WebApplicationFactory<Startup>` yourself in the constructor of `SaveTestDatabaseFixture`.

The reason is that we have very simplified lifetime support. We did not want to be in the business of resolving dependencies at the same level (i.e., collection fixture to collection fixture, class fixture to class fixture, etc.) because the rules can get out of hand quickly. You can only take dependencies on things from a wider scope. In v2, that means class fixtures can depend on collection fixtures; in v3, with the addition of assembly fixtures, that means class & collection fixtures can depend on assembly fixtures.

**Comment by James-Bryan** [comment]
Thanks for the confirmation Brad.
I will leave this discussion post open for now.
Once I have implemented the solution for this, I will document it here and close the ticket.

**Comment by James-Bryan** [comment]
The initial solution I have implemented is to:
- Continue using IClassFixture
- Only fully execute the fixture constructor the first time in, and keep a static reference to the fully initialized fixture
- Any additional instances of the fixture constructor exit early, and do not re-generate the test data
- All Test files reference the static fixture instead of the fixture passed into the test constructor
- The fixture still needs to be passed into the test constructor so the fixture constructor is fully executed when it is needed

Main Collection to provide the WebApplicationFactory
```
[CollectionDefinition("Tests")]
public class FixtureCollection : ICollectionFixture<WebApplicationFactory<Startup>>
{

}
```

Fixture uses the Injected WebApplicationFactory to initalize the test data
There is now a static property that keeps a reference to an instance of the fixture
If the static Instance has not been defined, run fixture initialization, and set the static instance to reference this instance of the fixture
If the static Instance has been defined, do not run fixture initialization
This ensures that the Database data is only generated once, no matter how many times the fixture is used
```
public class SaveTestDatabaseFixture : WebApplicationFactory<Startup>{

    public static SaveTestDatabaseFixture Instance;

    public TestData first_test_data { get; set; }

    public SaveTestDatabaseFixture(WebApplicationFactory<Startup> factory)
    {
          if (Instance != null)
          {
              return;
          }

          var context = factory.GetContext();
          ......
          first_test_data = new TestData();
          context.TestData.Add(first_test_data);
          context.SaveChanges();

          Instance = this;
    }
}
```

All tests use the static instance of the Database Fixture.
```
[Collection("Tests")]
public class TestDataTests : IClassFixture<SaveTestDatabaseFixture> {
    private readonly SaveTestDatabaseFixture databaseFixture;

    public TestDataTests (SaveTestDatabaseFixture _)
    {
        this.databaseFixture = SaveTestDatabaseFixture.Instance;
    }

    [Fact]
    public async Task Test1()
    {
        var x = databaseFixture.TestData.Id;
        //Run Test
    }
}

[Collection("Tests")]
public class TestData2Tests : IClassFixture<SaveTestDatabaseFixture> {
    private readonly SaveTestDatabaseFixture databaseFixture;

    public TestData2Tests (SaveTestDatabaseFixture _)
    {
        this.databaseFixture = SaveTestDatabaseFixture.Instance;
    }

    [Fact]
    public async Task Test2()
    {
        var x = databaseFixture.TestData2.Id;
        //Run Test
    }
}
```

As I have quite a few separate test files, this change reduced the amount of data being saved to the Test Database by 90% (2000 rows down to 200 rows).
It also reduced the execution time of the test suite (including Database setup time) from from 53 seconds to 20 seconds.

If there is a better/simpler way to implement this, let me know.
Otherwise, I think we can close this discussion.
